### PR TITLE
2.x HAL+JSON Formatting

### PR DESCRIPTION
**Pull request by e0ipso**
This will bring back HAL+JSON formatting.

When listing items with embedded items, this is what it looks like:
dhc 2015-04-15 17-31-42

When listing an item with embedded items, this is what it looks like:

``` javascript
{
    "hal:articles": [
        {
            "id": "88",
            "label": "Elit Euismod Sagaciter",
            "status": "1",
            "body": "<p>Antehabeo autem incassum. Adipiscing aliquip genitus jus luctus quibus.</p>",
            "_links": {
                "self": {
                    "href": "http://restful.local/api/v1.0/articles/88"
                }
            },
            "_embedded": {
                "hal:tags": [
                    {
                        "id": "3",
                        "label": "alpha",
                        "_links": {
                            "self": {
                                "href": "http://restful.local/api/v1.0/tags/3"
                            }
                        }
                    },
                    {
                        "id": "4",
                        "label": "beta"
                    }
                ]
            }
        }
    ]
}
```

### CODE REVIEW

**Comment by e0ipso** [comment]
@amitaibu this is ready for review.

Off-topic
@RoySegall I'm texting you in hangouts, I haven't seen you in all day.

File: src/Plugin/resource/Field/ResourceFieldBase.php [hunk 1]
```diff
@@ -244,4 +251,50 @@ public function isComputed() {
...
+  public function getMetadata($key) {
+    $path = explode(':', $key);
+    $leave = array_pop($path);
+    $element = $this->internalMetadataElement($key);
+
+    return isset($element[$leave]) ? $element[$leave] : NULL;
+  }
+
+  /**
+   * Returns the last array element from the nested namespace array.
```
Comment (amitaibu) [inline src/Plugin/resource/Field/ResourceFieldBase.php:277]: can you add more comments here - not trivial to follow the code flow.

**Comment by amitaibu** [comment]
Minor comment - after that can be merged

**Comment by mirzu** [comment]
This excites me.

**Comment by mirzu** [comment]
@thebruce take a look. . .

**Comment by e0ipso** [comment]
This is ready to re-review.

**Comment by e0ipso** [comment]
@mirzu @thebruce I changed the curie separator from `:` to `\`.

**Comment by mirzu** [comment]
According to the spec curies are separated by a colon ':'

I think the best thing to provide is a simple off switch. So that users that use a framework or templating language that has trouble with ':' in JSON keys can turn them off.

**Comment by e0ipso** [comment]
@mirzu I missed that from the specs. Since I moved the separator to a constant I will put i back as a `:`. Any formatter can override the constant value if they use a PHP version that supports lsb (which I believe is a pretty low version).

**Pull request merged**